File: c++/array/SetMatrix.cpp
class Solution {
public:
    void setZeroes(vector<vector<int>> &matrix) {
        bool firstRowZero = false;
        bool firstColZero = false;
        int rows = matrix.size();
        int cols = matrix[0].size();
        
        for (int i = 0; i < rows; i++) {
            if (matrix[i][0] == 0) {
                firstColZero = true;
                break;
            }
        }
        
        for (int i = 0; i < cols; i++) {
            if (matrix[0][i] == 0) {
                firstRowZero = true;
                break;
            }
        }
        
        for (int i = 1; i < rows; i++) {
            for (int j = 1; j < cols; j++) {
                if (matrix[i][j] == 0) {
                    matrix[i][0] = 0;
                    matrix[0][j] = 0;
                }
            }
        }
        
        for (int i = 1; i < rows; i++) {
            if (matrix[i][0] == 0) {
                for (int j = 1; j < cols; j++) {
                    matrix[i][j] = 0;
                }
            }
        }
        
        for (int i = 1; i < cols; i++) {
            if (matrix[0][i] == 0) {
                for (int j = 1; j < rows; j++) {
                    matrix[j][i] = 0;
                }
            }
        }
        
        if (firstRowZero) {
            for (int i = 0; i < cols; i++) {
                matrix[0][i] = 0;
            }
        }
        
        if (firstColZero) {
            for (int i = 0; i < rows; i++) {
                matrix[i][0] = 0;
            }
        }
        
        return;
    }
};

File: python/array/SetMatrix.py
class Solution:
    # @param matrix, a list of lists of integers
    # RETURN NOTHING, MODIFY matrix IN PLACE.
    def setZeroes(self, matrix):
        firstRowZero = False
        firstColZero = False
        rows = len(matrix)
        cols = len(matrix[0])
        
        for i in xrange(rows):
            if matrix[i][0] == 0:
                firstColZero = True
                break
        
        for i in xrange(cols):
            if matrix[0][i] == 0:
                firstRowZero = True
                break
            
        for i in xrange(1, rows):
            for j in xrange(1, cols):
                if matrix[i][j] == 0:
                    matrix[i][0] = 0
                    matrix[0][j] = 0
        
        for i in xrange(1, rows):
            if matrix[i][0] == 0:
                for j in xrange(1, cols):
                    matrix[i][j] = 0
                    
        for i in xrange(1, cols):
            if matrix[0][i] == 0:
                for j in xrange(1, rows):
                    matrix[j][i] = 0
        
        if firstRowZero:
            for i in xrange(cols):
                matrix[0][i] = 0
        
        if firstColZero:
            for i in xrange(rows):
                matrix[i][0] = 0
